File: src/main/java/io/github/mdsimmo/bomberman/messaging/Text.java
package io.github.mdsimmo.bomberman.messaging;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.util.HashMap;
import java.util.Map;

import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.YamlConfiguration;
import org.jetbrains.annotations.NotNull;

/**
 * A list of Contexteds that are defined in english.yml in the resource folder.
 */
public enum Text implements Contexted {

	MESSAGE_FORMAT ( "format.message" ),
	HEADING_FORMAT ( "format.heading" ),
	MAP_FORMAT ( "format.map" ),
	LIST_FORMAT ( "format.list" ),
	ITEM_FORMAT ( "format.item" ),

	GAME ( "word.game" ),
	ARENA ( "word.arena" ),

	TRUE ( "word.true" ),
	FALSE ( "word.false" ),

	EDIT_BUILD_DENIED ( "editmode.builddenied" ),
	EDIT_DESTROY_DENIED ( "editmode.destroydenied" ),

	TELEPORT_DENIED ( "game-play.teleport-denied" ),
	HIT_OPPONENT ( "game-play.hit-opponent" ),
	HIT_SUICIDE ( "game-play.hit-suicide" ),
	HIT_BY ( "game-play.hit-by" ),
	KILL_OPPONENT ( "game-play.kill-opponent" ),
	KILL_SUICIDE ( "game-play.kill-suicide" ),
	KILLED_BY ( "game-play.killed-by" ),
	PLAYER_KILLED( "game-play.player-killed" ),
	PLAYER_WON( "game-play.player-won" ),
	PLAYER_LEFT( "game-play.player-left" ),
	GAME_STARTING ( "game-play.starting" ),
	GAME_COUNT ( "game-play.count" ),
	GAME_STARTED( "game-play.started" ),
	COUNT_STOPPED ( "game-play.stopped-count" ),
	GAME_OVER ( "game-play.game-over" ),

	SUDDENDEATH_COUNT ( "suddendeath.count" ),
	SUDDENDEATH ( "suddendeath.start" ),
	TIMEOUT_COUNT ( "timeout.count" ),
	TIMEOUT ( "timeout.start" ),

	SCORE_ANNOUNCE ( "scores.announce" ),
	WINNERS_LIST ( "scores.winners" ),

	DENY_PERMISSION ( "command.deny-permission" ),
	INCORRECT_USAGE ( "command.incorrect-usage" ),
	UNKNOWN_COMMAND ( "command.unknown-command" ),
	MUST_BE_PLAYER ( "command.must-be-player" ),
	INVALID_NUMBER ( "command.invalid-number" ),
	INVALID_PLAYER ( "command.invalid-player" ),
	INVALID_MATERIAL ( "command.invalid-material" ),
	INVALID_SCHEMA ( "command.invalid-schema" ),
	INVALID_GAME ( "command.invalid-game" ),
	COMMAND_GROUP_USAGE("command.group.usage"),
	COMMAND_GROUP_EXAMPLE("command.group.example"),
	COMMAND_GROUP_EXTRA("command.group.extra"),
	COMMAND_FORMAT("command.format"),

	// no Bomberman name as it must not be changed
	BOMBERMAN_DESCRIPTION ( "command.bomberman.description" ),

	ARENA_CREATE_NAME ( "command.arenacreate.name" ),
	ARENA_CREATE_DESCRIPTION ( "command.arenacreate.description" ),
	ARENA_CREATE_USAGE ( "command.arenacreate.usage" ),
	ARENA_CREATE_EXAMPLE ( "command.arenacreate.example" ),
	ARENA_CREATE_EXTRA ( "command.arenacreate.extra" ),
	ARENA_CREATED ( "command.arenacreate.success" ),
	ARENA_CREATING ( "command.arenacreate.started" ),
	ARENA_NO_TARGET ( "command.arenacreate.notarget" ),
	ARENA_CREATE_TOO_BIG ( "command.arenacreate.too-big" ),
	ARENA_CREATE_VERY_SMALL ( "command.arenacreate.very-small" ),
	ARENA_CREATE_IN_USE ( "command.arenacreate.in-use" ),

	START_NAME ( "command.start.name" ),
	START_DESCRIPTION ( "command.start.description" ),
	START_USAGE ( "command.start.usage" ),
	START_EXAMPLE ( "command.start.example" ),
	START_EXTRA ( "command.start.extra" ),
	GAME_ALREADY_STARTED ( "command.start.already-started" ),
	GAME_START_SUCCESS ( "command.start.success" ),
	GAME_START_CANCELLED ( "command.start.cancelled" ),
	GAME_MORE_PLAYERS ( "command.start.more-players" ),

	STOP_NAME ( "command.stop.name" ),
	STOP_DESCRIPTION ( "command.stop.description" ),
	STOP_USAGE ( "command.stop.usage" ),
	STOP_EXAMPLE ( "command.stop.example" ),
	STOP_EXTRA ( "command.stop.extra" ),
	STOP_NOT_STARTED ( "command.stop.not-started" ),
	STOP_SUCCESS ( "command.stop.success" ),

	SET_SCHEMA_NAME ( "command.set-schema.name" ),
	SET_SCHEMA_DESCRIPTION ( "command.set-schema.description" ),
	SET_SCHEMA_USAGE ( "command.set-schema.usage" ),
	SET_SCHEMA_EXAMPLE ( "command.set-schema.example" ),
	SET_SCHEMA_EXTRA ( "command.set-schema.extra" ),
	SET_SCHEMA_GIP ( "command.set-schema.game-in-progress" ),
	SET_SCHEMA_STARTED ( "command.set-schema.started" ),
	SET_SCHEMA_SUCCESS ( "command.set-schema.success" ),

	AUTOSTART_NAME ( "command.autostart.name" ),
	AUTOSTART_DESCRIPTION ( "command.autostart.description" ),
	AUTOSTART_USAGE ( "command.autostart.usage" ),
	AUTOSTART_EXAMPLE ( "command.autostart.example" ),
	AUTOSTART_EXTRA ( "command.autostart.extra" ),
	AUTOSTART_ENABLED ( "command.autostart.enabled" ),
	AUTOSTART_DISABLED ( "command.autostart.disabled" ),

	STARTDELAY_NAME ( "command.start-delay.name" ),
	STARTDELAY_DESCRIPTION ( "command.start-delay.description" ),
	STARTDELAY_USAGE ( "command.start-delay.usage" ),
	STARTDELAY_EXAMPLE ( "command.start-delay.example" ),
	STARTDELAY_EXTRA ( "command.start-delay.extra" ),
	STARTDELAY_SET ( "command.start-delay.set" ),

	BOMBS_NAME ( "command.bombs.name" ),
	BOMBS_DESCRIPTION ( "command.bombs.description" ),
	BOMBS_USAGE ( "command.bombs.usage" ),
	BOMBS_EXAMPLE ( "command.bombs.example" ),
	BOMBS_EXTRA ( "command.bombs.extra" ),
	BOMBS_SET ( "command.bombs.set" ),

	FARE_NAME ( "command.fare.name" ),
	FARE_DESCRIPTION ( "command.fare.description" ),
	FARE_USAGE ( "command.fare.usage" ),
	FARE_EXAMPLE ( "command.fare.example" ),
	FARE_EXTRA ( "command.fare.extra" ),
	FARE_SET ( "command.fare.set" ),
	FARE_NONE ( "command.fare.none" ),
	FARE_XP ( "command.fare.xp" ),
	FARE_VAULT ( "command.fare.money" ),
	FARE_IN_HAND ( "command.fare.inhand" ),

	HANDICAP_NAME ( "command.handicap.name" ),
	HANDICAP_DESCRIPTION ( "command.handicap.description" ),
	HANDICAP_USAGE ( "command.handicap.usage" ),
	HANDICAP_EXAMPLE ( "command.handicap.example" ),
	HANDICAP_EXTRA ( "command.handicap.extra" ),
	HANDICAP_SET ( "command.handicap.set" ),
	
	LIVES_NAME ( "command.lives.name" ),
	LIVES_DESCRIPTION ( "command.lives.description" ),
	LIVES_USAGE ( "command.lives.usage" ),
	LIVES_EXAMPLE ( "command.lives.example" ),
	LIVES_EXTRA ( "command.lives.extra" ),
	LIVES_SET ( "command.lives.set" ),

	MINPLAYERS_NAME ( "command.minplayers.name" ),
	MINPLAYERS_DESCRIPTION ( "command.minplayers.description" ),
	MINPLAYERS_USAGE ( "command.minplayers.usage" ),
	MINPLAYERS_EXAMPLE ( "command.minplayers.example" ),
	MINPLAYERS_EXTRA ( "command.minplayers.extra" ),
	MINPLAYERS_SET ( "command.minplayers.set" ),
	MINPLAYERS_LESS_THAN_ONE ( "command.minplayers.more-than-zero" ),

	POWER_NAME ( "command.power.name" ),
	POWER_DESCRIPTION ( "command.power.description" ),
	POWERS_USAGE ( "command.power.usage" ),
	POWER_EXAMPLE ( "command.power.example" ),
	POWER_EXTRA ( "command.power.extra" ),
	POWER_SET ( "command.power.set" ),

	PRIZE_NAME ( "command.prize.name" ),
	PRIZE_DESCRIPTION ( "command.prize.description" ),
	PRIZE_USAGE ( "command.prize.usage" ),
	PRIZE_EXAMPLE ( "command.prize.example" ),
	PRIZE_EXTRA ( "command.prize.extra" ),
	PRIZE_POT ( "command.prize.pot" ),
	PRIZE_NONE ( "command.prize.none" ),
	PRIZE_XP ( "command.prize.xp" ),
	PRIZE_IN_HAND ( "command.prize.inhand" ),
	PRIZE_VAULT ( "command.prize.money" ),
	PRIZE_SET ( "command.prize.set" ),
			
	SET_NAME ( "command.set.name" ),
	SET_DESCRIPTION ( "command.set.description" ),

	SUDDENDEATH_NAME ( "command.suddendeath.name" ),
	SUDDENDEATH_DESCRIPTION ( "command.suddendeath.description" ),
	SUDDENDEATH_USAGE ( "command.suddendeath.usage" ),
	SUDDENDEATH_EXAMPLE ( "command.suddendeath.example" ),
	SUDDENDEATH_EXTRA ( "command.suddendeath.extra" ),
	SUDDENDEATH_OFF ( "command.suddendeath.off" ),
	SUDDENDEATH_REMOVED ( "command.suddendeath.removed" ),
	SUDDENDEATH_SET ( "command.suddendeath.set" ),

	TIMEOUT_NAME ( "command.timeout.name" ),
	TIMEOUT_DESCRIPTION ( "command.timeout.description" ),
	TIMEOUT_USAGE ( "command.timeout.usage" ),
	TIMEOUT_EXAMPLE ( "command.timeout.example" ),
	TIMEOUT_EXTRA ( "command.timeout.extra" ),
	TIMEOUT_OFF ( "command.timeout.off" ),
	TIMEOUT_REMOVED ( "command.timeout.removed" ),
	TIMEOUT_SET ( "command.timeout.set" ),

	CONVERT_NAME ( "command.convert.name" ),
	CONVERT_DESCRIPTION ( "command.convert.description" ),
	CONVERT_USAGE ( "command.convert.usage" ),
	CONVERT_EXAMPLE ( "command.convert.example" ),
	CONVERT_EXTRA ( "command.convert.extra" ),
	CONVERT_GAME_EXISTS ( "command.convert.game-exists" ),
	CONVERT_STARTED ( "command.convert.started" ),
	CONVERT_SUCCESS ( "command.convert.success" ),

	CREATE_NAME ( "command.create.name" ),
	CREATE_DESCRIPTION ( "command.create.description" ),
	CREATE_USAGE ( "command.create.usage" ),
	CREATE_EXAMPLE ( "command.create.example" ),
	CREATE_EXTRA ( "command.create.extra" ),
	CREATE_GAME_EXISTS ( "command.create.game-exists" ),
	CREATE_NEED_SELECTION ( "command.create.need-selection" ),
	CREATE_STARTED ( "command.create.started" ),
	CREATE_SUCCESS ( "command.create.success" ),
	CREATE_DEFAULTMISSING ( "command.create.default-missing" ),
	CREATE_SCHEMA_NOT_FOUND ( "command.create.schema-not-found" ),

	DESTROY_NAME ( "command.destroy.name" ),
	DESTROY_DESCRIPTION ( "command.destroy.description" ),
	DESTROY_USAGE ( "command.destroy.usage" ),
	DESTROY_EXAMPLE ( "command.destroy.example" ),
	DESTROY_EXTRA ( "command.destroy.extra" ),
	DESTROY_SUCCESS ( "command.destroy.success" ),

	FORGET_NAME ( "command.forget.name" ),
	FORGET_DESCRIPTION ( "command.forget.description" ),
	FORGET_USAGE ( "command.forget.usage" ),
	FORGET_EXAMPLE ( "command.forget.example" ),
	FORGET_EXTRA ( "command.forget.extra" ),
	FORGET_SUCCESS ( "command.forget.success" ),

	GAME_NAME ( "command.game.name" ),
	GAME_DESCRIPTION ( "command.game.description" ),

	GAMELIST_NAME ( "command.gamelist.name" ),
	GAMELIST_DESCRIPTION ( "command.gamelist.description" ),
	GAMELIST_USAGE ( "command.gamelist.usage" ),
	GAMELIST_EXAMPLE ( "command.gamelist.example" ),
	GAMELIST_EXTRA ( "command.gamelist.extra" ),
	GAMELIST_GAMES ( "command.gamelist.games" ),

	INFO_NAME ( "command.info.name" ),
	INFO_DESCRIPTION ( "command.info.description" ),
	INFO_USAGE ( "command.info.usage" ),
	INFO_EXAMPLE ( "command.info.example" ),
	INFO_EXTRA ( "command.info.extra" ),
	INFO_DETAILS ( "command.info.details" ),


	JOIN_NAME ( "command.join.name" ),
	JOIN_DESCRIPTION ( "command.join.description" ),
	JOIN_USAGE ( "command.join.usage" ),
	JOIN_EXAMPLE ( "command.join.example" ),
	JOIN_EXTRA ( "command.join.extra" ),
	JOIN_GAME_STARTED ( "command.join.game-started" ),
	JOIN_ALREADY_JOINED ( "command.join.already-joined" ),
	GAME_FULL ( "command.join.game-full" ),
	TOO_POOR ( "command.join.poor-man" ),
	CANT_JOIN ( "command.join.cant-join" ),
	PLAYER_JOINED ( "command.join.player-joined" ),

	LEAVE_NAME ( "command.leave.name" ),
	LEAVE_DESCRIPTION ( "command.leave.description" ),
	LEAVE_USAGE ( "command.leave.usage" ),
	LEAVE_EXAMPLE ( "command.leave.example" ),
	LEAVE_EXTRA ( "command.leave.extra" ),
	LEAVE_SUCCESS ( "command.leave.success" ),
	LEAVE_NOT_JOINED ( "command.leave.not-joined" );

	private final String text;
	
	Text( String path ) {
		this.text = YAMLLanguage.lang.getString( path );
		
		// any error's here should always get caught during development
		if ( text == null )
			throw new RuntimeException( "No default message for text: " + path  );
	}

	@Override
	public Contexted with(String key, Formattable thing) {
		return new Contexted() {
			Map<String, Formattable> things = new HashMap<>();

			@Override
			public Contexted with(String key, Formattable arg) {
				things.put(key, arg);
				return this;
			}

			@NotNull
			@Override
			public Message format() {
				return Expander.expand(text, things);
			}
		}.with(key, thing);
	}

	@Override
	public Message format() {
		return Expander.expand(text, Map.of());
	}

	// class to read the English language text from
	static class YAMLLanguage {
		static final YamlConfiguration lang;
		static {
			InputStream in = Text.class.getClassLoader().getResourceAsStream("english.yml");
			Reader reader = new BufferedReader( new InputStreamReader( in ) );
			lang = new YamlConfiguration();
			try {
				lang.load( reader );
			} catch ( IOException | InvalidConfigurationException e ) {
				e.printStackTrace();
			}
		}
	}

}

File: src/main/java/io/github/mdsimmo/bomberman/messaging/Text.kt
package io.github.mdsimmo.bomberman.messaging

import io.github.mdsimmo.bomberman.Bomberman
import org.bukkit.configuration.InvalidConfigurationException
import org.bukkit.configuration.file.YamlConfiguration
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.Reader
import java.nio.file.Files
import kotlin.io.path.exists

/**
 * A list of Contexteds that are defined in messages.yml. If messages.yml has not been created, then it will default to
 * default_messages.yml in the resource folder.
 */
enum class Text(path: String) : Formattable {
    MESSAGE_FORMAT("format.message"),
    ITEM_FORMAT("format.item"),
    PLAYER_WON("game-play.player-won"),
    GAME_COUNT("game-play.count"),
    GAME_STARTED("game-play.started"),
    DENY_PERMISSION("command.deny-permission"),
    INCORRECT_USAGE("command.incorrect-usage"),
    UNKNOWN_COMMAND("command.unknown-command"),
    MUST_BE_PLAYER("command.must-be-player"),
    INVALID_NUMBER("command.invalid-number"),
    INVALID_TARGET_SELECTOR("command.invalid-target-selector"),
    COMMAND_GROUP_HELP("command.group.help"),
    COMMAND_GROUP_USAGE("command.group.usage"),
    COMMAND_GROUP_EXAMPLE("command.group.example"),
    COMMAND_GROUP_EXTRA("command.group.extra"),
    COMMAND_HELP("command.help"),
    COMMAND_CANCELLED("command.cancelled"),
    BOMBERMAN_DESCRIPTION("command.bomberman.description"),
    START_NAME("command.start.name"),
    START_DESCRIPTION("command.start.description"),
    START_USAGE("command.start.usage"),
    START_EXAMPLE("command.start.example"),
    START_EXTRA("command.start.extra"),
    START_FLAG_OVERRIDE_DESC("command.start.flags.o.description"),
    START_FLAG_DELAY_DESC("command.start.flags.d.description"),
    START_FLAG_DELAY_EXT("command.start.flags.d.ext"),
    GAME_ALREADY_STARTED("command.start.already-started"),
    GAME_ALREADY_COUNTING("command.start.already-counting"),
    GAME_NO_PLAYERS("command.start.no-players"),
    GAME_START_SUCCESS("command.start.success"),
    STOP_NAME("command.stop.name"),
    STOP_DESCRIPTION("command.stop.description"),
    STOP_USAGE("command.stop.usage"),
    STOP_EXAMPLE("command.stop.example"),
    STOP_EXTRA("command.stop.extra"),
    STOP_NOT_STARTED("command.stop.not-started"),
    STOP_SUCCESS("command.stop.success"),
    STOP_TIMER_STOPPED("command.stop.timer-stopped"),
    RELOAD_NAME("command.reload.name"),
    RELOAD_DESCRIPTION("command.reload.description"),
    RELOAD_USAGE("command.reload.usage"),
    RELOAD_EXAMPLE("command.reload.example"),
    RELOAD_EXTRA("command.reload.extra"),
    RELOAD_SUCCESS("command.reload.success"),
    RELOAD_CANNOT_LOAD("command.reload.cannot-load"),
    CONFIGURE_NAME("command.configure.name"),
    CONFIGURE_DESCRIPTION("command.configure.description"),
    CONFIGURE_USAGE("command.configure.usage"),
    CONFIGURE_EXAMPLE("command.configure.example"),
    CONFIGURE_EXTRA("command.configure.extra"),
    CONFIGURE_PROMPT_CREATIVE("command.configure.prompt-creative"),
    CONFIGURE_TITLE_MAIN("command.configure.title.main"),
    CONFIGURE_TITLE_GENERAL("command.configure.title.general"),
    CONFIGURE_TITLE_BLOCKS("command.configure.title.blocks"),
    CONFIGURE_TITLE_LOOT("command.configure.title.loot"),
    CONFIGURE_TITLE_INVENTORY("command.configure.title.inventory"),
    CONFIGURE_BACK("command.configure.back"),
    CONFIGURE_DESTRUCTIBLE("command.configure.blocks.destructible.name"),
    CONFIGURE_DESTRUCTIBLE_DESC("command.configure.blocks.destructible.description"),
    CONFIGURE_INDESTRUCTIBLE("command.configure.blocks.indestructible.name"),
    CONFIGURE_INDESTRUCTIBLE_DESC("command.configure.blocks.indestructible.description"),
    CONFIGURE_PASS_KEEP("command.configure.blocks.pass-keep.name"),
    CONFIGURE_PASS_KEEP_DESC("command.configure.blocks.pass-keep.description"),
    CONFIGURE_PASS_DESTROY("command.configure.blocks.pass-destroy.name"),
    CONFIGURE_PASS_DESTROY_DESC("command.configure.blocks.pass-destroy.description"),
    CONFIGURE_LIVES("command.configure.general.lives"),
    CONFIGURE_FUSE_TICKS("command.configure.general.fuse-ticks"),
    CONFIGURE_FIRE_TICKS("command.configure.general.fire-ticks"),
    CONFIGURE_IMMUNITY_TICKS("command.configure.general.immunity-ticks"),
    CONFIGURE_TNT_BLOCK("command.configure.general.tnt-block"),
    CONFIGURE_FIRE_ITEM("command.configure.general.fire-item"),
    CONFIGURE_LOOT_SLOT("command.configure.loot.slot"),
    CONFIGURE_LOOT_BLOCK("command.configure.loot.block"),
    CONFIGURE_LOOT_ITEM("command.configure.loot.item"),
    CONFIGURE_LOOT_WEIGHT("command.configure.loot.weight"),
    CREATE_NAME("command.create.name"),
    CREATE_DESCRIPTION("command.create.description"),
    CREATE_USAGE("command.create.usage"),
    CREATE_EXAMPLE("command.create.example"),
    CREATE_EXTRA("command.create.extra"),
    CREATE_GAME_EXISTS("command.create.game-exists"),
    CREATE_GAME_FILE_CONFLICT("command.create.file-conflict"),
    CREATE_GAME_FILE_NOT_FOUND("command.create.file-not-found"),
    CREATE_NEED_SELECTION("command.create.need-selection"),
    CREATE_SUCCESS("command.create.success"),
    CREATE_ERROR("command.create.error"),
    CREATE_SCHEMA_NOT_FOUND("command.create.schema-not-found"),
    CREATE_FLAG_SCHEMA("command.create.flags.s.description"),
    CREATE_FLAG_GAME("command.create.flags.g.description"),
    CREATE_FLAG_TEMPLATE("command.create.flags.t.description"),
    CREATE_FLAG_WAND("command.create.flags.w.description"),
    CREATE_FLAG_SCHEMA_EXT("command.create.flags.s.ext"),
    CREATE_FLAG_GAME_EXT("command.create.flags.t.ext"),
    CREATE_FLAG_TEMPLATE_EXT("command.create.flags.g.ext"),
    DELETE_NAME("command.delete.name"),
    DELETE_DESCRIPTION("command.delete.description"),
    DELETE_USAGE("command.delete.usage"),
    DELETE_EXAMPLE("command.delete.example"),
    DELETE_EXTRA("command.delete.extra"),
    DELETE_SUCCESS("command.delete.success"),
    UNDO_NAME("command.undo.name"),
    UNDO_DESCRIPTION("command.undo.description"),
    UNDO_USAGE("command.undo.usage"),
    UNDO_EXAMPLE("command.undo.example"),
    UNDO_EXTRA("command.undo.extra"),
    UNDO_DELETED("command.undo.deleted"),
    UNDO_SUCCESS("command.undo.success"),
    UNDO_UNKNOWN_GAME("command.undo.unknown-game"),
    GAMELIST_NAME("command.list.name"),
    GAMELIST_DESCRIPTION("command.list.description"),
    GAMELIST_USAGE("command.list.usage"),
    GAMELIST_EXAMPLE("command.list.example"),
    GAMELIST_EXTRA("command.list.extra"),
    GAMELIST_GAMES("command.list.games"),
    INFO_NAME("command.info.name"),
    INFO_DESCRIPTION("command.info.description"),
    INFO_USAGE("command.info.usage"),
    INFO_EXAMPLE("command.info.example"),
    INFO_EXTRA("command.info.extra"),
    INFO_DETAILS("command.info.details"),
    JOIN_NAME("command.join.name"),
    JOIN_DESCRIPTION("command.join.description"),
    JOIN_USAGE("command.join.usage"),
    JOIN_EXAMPLE("command.join.example"),
    JOIN_EXTRA("command.join.extra"),
    JOIN_FLAG_TARGET("command.join.flags.t.description"),
    JOIN_FLAG_TARGET_EXT("command.join.flags.t.ext"),
    JOIN_GAME_STARTED("command.join.game-started"),
    JOIN_ALREADY_JOINED("command.join.already-joined"),
    JOIN_GAME_FULL("command.join.game-full"),
    JOIN_SUCCESS("command.join.success"),
    LEAVE_NAME("command.leave.name"),
    LEAVE_DESCRIPTION("command.leave.description"),
    LEAVE_USAGE("command.leave.usage"),
    LEAVE_EXAMPLE("command.leave.example"),
    LEAVE_EXTRA("command.leave.extra"),
    LEAVE_FLAG_TARGET("command.leave.flags.t.description"),
    LEAVE_FLAG_TARGET_EXT("command.leave.flags.t.ext"),
    LEAVE_SUCCESS("command.leave.success"),
    LEAVE_NOT_JOINED("command.leave.not-joined");

    private val text: String

    init {
        // any error's here should always get caught during development
        text = YAMLLanguage.server?.getString(path)
                ?: YAMLLanguage.builtin.getString(path)
                ?: throw RuntimeException("No default message for text: $path")
    }

    override fun applyModifier(arg: Message): Formattable {
        return this
    }

    override fun format(context: Context): Message {
        return Expander.expand(text, context.plus(YAMLLanguage.textContext))
    }

    // class to read the English language text from
    private object YAMLLanguage {
        val builtin: YamlConfiguration
        val server: YamlConfiguration?
        val textContext: Context

        init {
            val input = Text::class.java.classLoader.getResourceAsStream("default_messages.yml")!!
            val reader: Reader =
                    BufferedReader(InputStreamReader(input))
            builtin = YamlConfiguration()
            server = YamlConfiguration()
            try {
                builtin.load(reader)
                // Instance may be null when testing
                Bomberman.instance?.let { plugin ->
                    val custom = plugin.language()
                    if (custom.exists())
                        Files.newBufferedReader(custom).use { server.load(it) }
                }
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (e: InvalidConfigurationException) {
                e.printStackTrace()
            }
            textContext = Context(true).addFunctions {
                server.getString(it) ?: builtin.getString(it)
            }
        }
    }
}